Implementation of the Refinery

legionth opened the pull request:
This is the implementation for the `Call for Bids - ILIAS Security, Refinery Interface` and the according submission.

## Why is this PR so big?

The change of 185 Files is very hard to read. We understand this.

This PR contains the Interfaces and concrete implementation of the new `Refinery` library, but also the transfer of the libraries `Validation` and `Transformation`(including the needed namespace changes in the consumer code).

If desired I can extract the transition of the `Validation` and `Transformation`, so this transition can be reviewed first. The move of this libraries is needed _first_, because the new Refinery components have dependencies to these libraries.

In our internal review process we used *very small and atomic* pull requests (~50). This would also be an option to ease the review process.

## README

The [README.md](/src/Refinery/README.md) should contain every information, to work with this library.

## Unittests

As known the Unittest can be executed on a local copy of this repository via:

```
$ phpunit tests/Refinery/
```

## Examples

Similar to the examples of the [UI library](src/UI/examples/) these are mostly for show cases and only operable with an initialized ILIAS (also described in the [README](/src/Refinery/README.md) )

Commit:
Move Transformation Library into Refinery Library

Commit:
Change namespace of Transformation sub-library to be working again

Commit:
Move 'Validation' library to 'Refinery' library

Commit:
Rename Validation namespace to 'ILIAS\Refinery\Validation'

Commit:
Add README for Refinery

Commit:
Add apply to method to Transformation Interface

Commit:
Correct Custom Transformation Behavior

Commit:
Add Untitests to test new applyTo-method

Commit:
Rename Validation method applyTo to restrictTo

Commit:
Add refinery interfaces for upcoming implementation

Commit:
Complete implementation of the tuple transformation

Commit:
Parallel and Series Transformation implementation

Commit:
Add basic tests for factory, groups and primitve transformation

Commit:
Complete ListTransformation implementation

Commit:
Complete ListTransformation implementation

Commit:
Add DictionaryTransformation Implementation

Commit:
Complete implemenataion new object transformation

Commit:
Complete implementation of NewMethodTransformation

Commit:
Add RecordTransformation Implementation

Commit:
Add a alphanumeric data type

Commit:
Add FloatRange data type

Commit:
Use direct dependency injection

Commit:
Primitive type tranformation

Commit:
Use InvalidArgumentException instead of ilException

Commit:
Strict type declaration

Commit:
Strict type declaration

Commit:
Strict type declaration

Commit:
Throw InvalidArgumentException instead of ilException

Commit:
Add StrictFloatRange that does NOT allow minimum and maximum as equal

Commit:
Add implementation for Integer Ranges

Commit:
Add StrictIntegerRange that does NOT allow minimum and maximum to be equal

Commit:
Add PositveInteger datatype

Commit:
Stric type check in primitive transformation classes

Commit:
Adapt Series and Parallel according to strict  condition

Commit:
Define basic kindlyTo group interface

Commit:
Update README

Commit:
Declare Strict Types

Commit:
Declare strict types

Commit:
Declare Strict Types

Commit:
Declare strict types

Commit:
Add new contstrain to check if the array contains only elements of the same type

Commit:
Remove unsused parameter

Commit:
Check for same classes in array

Commit:
Add type hints

Commit:
Change exception type to InvalidArgumentException

Commit:
Add additional tests for private methods and exception

Commit:
Fix: Dictonary unit test, tests for wrong exception

Commit:
Add checs for type error and exceptions in constructor

Commit:
Add check for error

Commit:
Update README

Commit:
Add implementation of In\Group

Commit:
Adapt interface

Commit:
Use streamlined validation for dictionary implementation

Commit:
Move transformation tests into subfolder

Commit:
Add implementation of To/Group

Commit:
Add tests for parameters in newMethod transformations

Commit:
Adapt namespace of In\BasicGroup

Commit:
Implementation of Refinery/Factory interface

Commit:
Add primitive KindlyToTransformations

Commit:
Implementation of KindlyTo/DictionaryImplementation

Commit:
Add tests for objects and arrays

Commit:
Implementation of KindlyTo/RecordOf transformation

Commit:
Implementation of KindlyTo/TupleOf transformation

Commit:
Fixes

Commit:
Fixx tuple transformation

Commit:
Add Implementation of KindlyTo/BasicGroup

Commit:
Add kindlyTo Implementations to factory

Commit:
Add examples for In/Group

Commit:
Add examples for To/Group

Commit:
Add examples for KindlyTo/Group

Commit:
Update README

Commit:
Change Namespace BasicFactoryTest

Commit:
Change Namespace In/BasicGroupTest

Commit:
Change Namespace In/ParallelTest

Commit:
Change Namespace In/SeriesTest

Commit:
Change Namespace KindlyTo/BasicGroupTest

Commit:
Change Namespace KindlyTo/BooleanTransformationTest

Commit:
Change Namespace KindlyTo/DictionaryTransformationTest

Commit:
Change Namespace KindlyTo/FloatTransformationTest

Commit:
Change Namespace KindlyTo/IntegerTransformationTest

Commit:
Change Namespace KindlyTo/ListTransformationTest

Commit:
Change Namespace KindlyTo/RecordTransformationTest

Commit:
Change Namespace KindlyTo/StringTransformationTest

Commit:
Change Namespace KindlyTo/TupleTransformationTest

Commit:
Change Namespace To/BasicGroupTest

Commit:
Change Namespace To/BooleanTransformationTest

Commit:
Change Namespace To/DictionaryTransformationTest

Commit:
Change Namespace To/FloatTransformationTest

Commit:
Change Namespace To/IntegerTransformationTest

Commit:
Change Namespace To/ListTransformationTest

Commit:
Change Namespace To/NewMethodTransformationTest

Commit:
Change Namespace To/NewObjectTransformationTest

Commit:
Change Namespace To/RecordTransformationTest

Commit:
Change Namespace To/StringTransformationTest

Commit:
Change Namespace To/TupleTransformationTest

Commit:
Add assertions in example functions

Commit:
Add assertions to examples

Commit:
Add assertions to examples

Commit:
Fix: Wrong type declaration in method declaration

Commit:
Add BasicFactory of Refinery to DIC

Commit:
Add strict type declaration for refinery class files

Commit:
Examples have wrong values

Commit:
Move whole instantiation of factories in callback function

Commit:
Fix KindlyTo/TupleTransformation

Commit:
Fix To/TupleTransformation

Commit:
Add common base classe to easier replace extension of PHPUnit

mjansenDatabay committed:
Revert "Merge branch 'kindly-to-examples' into 'trunk'"

This reverts merge request !42

Commit:
Fix tuple example

Commit:
Add examples for KindlyTo

Commit:
Add maintenance.json to refinery library

Commit:
Update READEM

Commit:
Add CONTRIBUTING.md

Commit:
Add quickstart example to README

Commit:
Replace maintainers wih coordinators

Commit:
Adaptions to CONTRIBUTING.md

Commit:
Update maintenance.json

Commit:
Additions to the text

Commit:
Add description for possible Jour Fixe discussions

Commit:
Remove CoC + more service specific PR description

klees commented on src/Refinery/README.md:
```diff
@@ -0,0 +1,332 @@
+
+$result = $transformation->transform(3.5); // Will throw exception because, values is not an integer value
+$result = $transformation->transform('hello'); // Will throw exception because, values is not an integer value
+$result = $transformation->transform(3); // $result = 10
```
```suggestion
$result = $transformation->transform(3); // $result = 3
```

klees commented on src/Refinery/examples/To/06-to-new-object.php:
```diff
@@ -0,0 +1,43 @@
+
+	$result = $transformation->transform(array('firstParameter', 2, 'thirdParameter'));
+
+	return assert('firstParameter' === $result);
```
```suggestion
	return assert('firstParameter' === $result->say());
```

klees commented on src/Data/StrictIntegerRange.php:
```diff
@@ -0,0 +1,49 @@
+class StrictIntegerRange
+{
+	/**
+	 * @var Integer
```
```suggestion
	 * @var IntegerRange
```

klees commented on src/Data/StrictIntegerRange.php:
```diff
@@ -0,0 +1,49 @@
+namespace ILIAS\Data;
+
+
+use phpDocumentor\Reflection\Types\Integer;
```
```suggestion
```

klees commented on src/Data/StrictIntegerRange.php:
```diff
@@ -0,0 +1,49 @@
+	public function __construct(int $minimum, int $maximum)
+	{
+		if ($minimum === $maximum) {
+			throw new \InvalidArgumentException(sprintf('The minimum value("%s") is not a integer', $minimum));
```
```suggestion
			throw new \InvalidArgumentException(sprintf('The maximum("%s") can NOT be same than the minimum("%s")', $maximum, $minimum));
```

klees commented on src/Refinery/BasicFactory.php:
```diff
@@ -0,0 +1,68 @@
+	private $validationFacory;
+
+	/**
+	 * @param \ilLanguage|null $language
```
```suggestion
```

klees commented on src/Refinery/CONTRIBUTING.md:
```diff
@@ -0,0 +1,61 @@
+
+When contributing to this service,
+please first discuss the change you wish to make via
+issue, email, or any other method with the owners of
```
```suggestion
issue, email, or any other method with the coordinators of
```

klees commented on src/Refinery/CONTRIBUTING.md:
```diff
@@ -0,0 +1,61 @@
+
+Bug fixes and improvements can always be entered via a Pull Request.
+
+A fix that redefines that behaviour of an implementation are considered
```
```suggestion
A fix that redefines the intended behaviour of an implementation are considered
```
Every fix changes some behaviour, we would not need them otherwise. So, in the end, every fix might break something.

klees commented on src/Refinery/CONTRIBUTING.md:
```diff
@@ -0,0 +1,61 @@
+## Breaking Changes
+
+A change that is not backwards compatible must be
+discussed with the [Jour Fixe](https://docu.ilias.de/goto_docu_wiki_wpage_391_1357.html).
```
```suggestion
announced one month in advance on the [Jour Fixe](https://docu.ilias.de/goto_docu_wiki_wpage_391_1357.html).
```

See 10. [here](https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-4/src).

klees reviewed (changes requested):
Hello @legionth!

Thank you a lot for implementing this important piece. I'm really looking forward to see this in action.

This already looks good, but, as most probably expected for such a huge and important changeset, I have questions, suggestions and change requests. Before I go into details, I especially like to point out three issues that I deem crucial for the success of the library:

* I'm currently missing `ConstraintViolation` (which might better be called `ConstraintViolationException`) in this PR (see (initial concept paper)[https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/input-processing.md]). This exception will be crucial to produce human readable error messages with i18n without binding everything to `ilLng`.
* We need to have a close look into the basic transformations behind `kindlyTo` and `to`. While the "kindlyTo"-variety was supposed to be used in situations where it is advisable to follow [Postels Law](https://en.wikipedia.org/wiki/Robustness_principle), the stricter "to"-variety is supposed to be used in situations, where we exactly know the shape of data that is incoming. This both is about the incoming data. We should be able to have the same expectations for the data after the
transformations in both varieties. This seems to be implemented slightly different currently, but I might be missing something. Also, the initial paper wasn't very clear, how we exactly could implement the two variants for the different transformations. These transformations will be crucial for the usability of the library as well as for the safety, so we should be diligent here.
* The move to merge the two previous libraries is not just some organisational move but the result of a deeper connection between the concepts encoded in the libs. That is: A `Transformation` and a `Constraint` basically are the same thing. They turn a `Result` into another `Result`. A `Constraint` is only special in the regard, that it does nothing to the value-part of `Result`, but might turn a `Result` that `isOk` into one that `isError` if the value does not fit. Both are just part of some processing pipeline `Result` -> `Result` -> `Result` -> ... There are `Transformations` that do something to values as well as constrain them, like `kindlyTo\IntegerTransformation`. This to me implies, overall, that we should drastically deemphasize the difference between `Transformation` and `Constraint`, because it is just not very important.

Now, the gory details.

Please answer the following questions:
- [ ] Why don't we use `string` for the [`$value` in `Alphanumeric`](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-3516fca3958b8f9aef0f542168e7582dR18)? Users of `Alphanumeric::getValue` would either break some typehint, examine the value or use `asString` anyway, because they cannot possibly know, how the instance at hand was constructed. So we could store (and return) a string without any loss and remove the `asString`-method.
- [x] Why don't we introduce `*Range::spans` to check if a given number is within the range?
- [x] Why did you introduce these `*AsInteger`-methods (etc.) for the `*Range` types? Why not just use `minimum` and `maximum` and let the type hinting talk about the type?
- [x] This might be the same question as above, but: Why does `PositiveInteger` use `asInteger` instead of `getValue`? Why does Alphanumeric have both kinds of methods?
- [x] Did you consciously use `issue` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-efbd1af760ce16b87196f3735b2e7253R5)? This would mean that we use GitHub-issues, right? In general I'm open to that, but I also think we need to consider this move carefully.
- [x] Why do you use `ilException` sometimes (e.g. [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-91bb549b5dcea87f9ae5e73707e61175R40))?
- [x] Do you have any idea how the examples in files and the READMEs could be tested systematically? We have a similar challenge in the UI-framework and will have it in other places if we create more documentation for developers.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Use `if (!preg_match(...))` [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-3516fca3958b8f9aef0f542168e7582dR28). This is what the return-value of `preg_match` is made for, while the behaviour of `$matches` is undocumented if the string did not match.
- [x] Get rid of the `Factory`-interfaces and let the implementations be the real deal (i.e. remove `Basic`). Although you used the style we use in the UI-framework, I cannot see that we gain much from that style here. According to the [rules of the ./src-folder](https://github.com/ILIAS-eLearning/ILIAS/tree/release_5-4/src), users are allowed to consider both, the interface and the implementation, to be public. Also, the interfaces don't seem to hide much, there are no internal-but-public methods on the implementation. I currently wonder if hiding factories behind interfaces ever makes sense and even would question that choice for the UI-framework. Thoughts?
- [x] Implement two traits `DeriveTransformFromApplyTo` and `DeriveApplyToFromTransform`, that do as they say. You could use them to get rid of either of the two methods in your implementations and thus avoid differences in the implementations. This would also make debugging easier and get rid of some changes.
- Lets get rid of the two sub-libraries `Validation` and `Transformation` completely. This would imo mean:
    * [ ] Move `Constraint` and `Transformation` to the top-level of the `Refinery`.
    * [ ] Find a place for the existing constraints and transformations in the new factory hierarchy. They would most likely need to be placed under the (currently missing) groups for native and data types. Remove duplicates when doing so. `IsString`, e.g., would be the same as `To\StringTransformation` imo.
    * [ ] Get rid of the two factories from `Validation` and `Constraints` and only use the `Refinery\Factory`.
    * [ ] Stop talking about the two libraries in the README and possibly other places. Keep the reference to these libraries as a historical view in the docu for input processing.
- [x] Use `call_user_func_array` in `NewMethodTransformation::transform`.

Please implement the following changes:
- [x] Please indicate if the `Range`-types are supposed to include or exclude the minimum and maximum values. I would further suggest to use the mathematical name `Interval` for these types. The question above then could be answered by choosing one of the canonical names "open interval" or "closed interval" (or even create a type for both).
- [x] Add the new data-types to the factory of the data-library.
- [x] I'm missing some methods on the factory with regards to the [original paper that spawned this implementation](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/input-processing.md): `identity`, `allOf`, `byTrying`, `optional`, `customTransformation` and `customConstraint`, as well as groups for types in data (as required/available). Is this intentional?
- [x] Remove this [if](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-8ece6df51f003fbb84891c8d04946fbaR64). This would only be the case if the user provided no transformations. An empty array as result would then be correct. That `if` on the other hand, would create a `Result` in an `array` in a `Result`, which would be odd. `transform` behaves exactly like this. This could be made simple by using a `DeriveApplyToFromTransform`-trait as suggested above.
- [x] Fix `Series::applyTo` (and add a unit-test for it). `transform` should behave correct. `applyTo` will return a double-wrapped value, because `$result` already is an instance of `Result`. The `foreach`, on the other hand, will do nothing, because `$resultObject` is never used. This could be made simple by using a `DeriveApplyToFromTransform`-trait as suggested above.
- [x] The basic transformations for `bool`, `float` and `integer` need some more accurate implementation, in both variants (`to` and `kindlyTo`). Note, that this is crucial for safety as well as for usability, as these transformations will appear in almost every chain of transformations that our consumers will build. There might be a missunderstanding on why and how we would want to use the
varieties of transformations. This all might be to much to discuss here, though, we might need some VC for that. Please initiate one, if you also feel that a VC would help. I write down what I have noticed so far:
    * `KindlyTo\BooleanTransformation` uses a cast to bool. This might be to on the one hand. Would we really want to coerce missing values (`""` or `null`) to `false`? This might be little on the other hand. Would we want to consider `"true"` as `true` but `"false"` as `true` as well? 
    * `KindlyTo\FloatTransformation` uses a cast to float. Would we want to consider `"1.2 Headline of this Paragraph in a text" to be the float `1.2`?
    * `KindlyTo\IntegerTransformation` uses a cast to integer. Would we want to consider the float `1.999999999` to be the integer `1`? I would refuse to discard that information here and simply not make the rounding operation that implicitely. We might want some rounding-transformations for that purpose. Also: I cannot see why we would want to interpret `"hello"` as `0`, even when being kind with the incoming data.
    * `KindlyTo\RecordTransformation` asserts, that the delivered array has the same size than the requested array. For me it would be sufficient if at least the requested keys would be present for the kind-variety of the transformation. If there are more keys in the delivered array I would just discard them. Same thing for `KindlyTo\TupleTransformation`.
    * `KindlyTo\StringTransformation` could well cast objects to strings if they had a `__toString`-method.
    * `To\DictionaryTransformation` should not check if the transformed and the original value are the same. I currently cannot see how this needs to be different from `KindlyTo\DictionaryTransformation`. Same goes for `To\ListTransformation` and `To\TupleTransformation`.
- [x] Remove the check with `IsArrayOfSameType` in `To\DictionaryTransformation`, `To\ListTransformation` and `To\TupleTransformation`. For the first to cases it imo would be perfectly sensible to have a list where in each location we, e.g., either have an int or a float, if the transformation we want to be applied produces these. For a tuple this even might be more normal if we follow the
definition of tuple from the input-paper (basically: record where keys are indexed with integers).
- [x] Only use the factory in the README instead of directly constructing the transformations. People are supposed to use the factory, so lets deliver the according templates for copy and paste.
- [x] In the [Quickstart Example](https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-59d18120fda5d1f83df62e314480332bR41) of the README, use the data-factory to create `Alphanumeric` by directly attaching it to the chain of transformations via `to->data`.
- [x] `To\Group::data` should return a transformation that creates a type from `ILIAS\Data\, not the original factory of `ILIAS\Data`. That is, it should be a thin wrapper around `NewMethodTransformation` where the factory is the data-factory and the method-name needs to be passed to `To\Group::data` by the consumer.
- [ ] Make `Validation\Constraint` extend from `Transformation\Transformation`.
- [x] Implement `ConstraintViolation` as outlined in the [concept paper](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/input-processing.md) and use it for your datatypes and examples. This will be a crucial part to connect data types and the Constraint-machinery.

I think the README needs some more polish, but this is hard to do via actionable change-requests. In general, I would try to give more examples from the "How and why would I do XYZ"-perspective, explain some core concepts (transformation, constraint, to vs. kindlyTo, composition: in-series/in-parallel/listOf/recordOf/...), then reference deeper locations like the factory to avoid the duplication of
information. Writing is hard, especially technical documentation, so we might wait with the README after the implementation is finished or even find (convince/force/...) someone else to write that stuff, if you do not feel to be up to that task.

Finally: This kind of big reviews is what you get for PRs this big =) But jokes aside: It took me some time to review this thing and write the text, but it certainly was worth it regarding the importance of the topic and the change set. I cannot see how I should have been faster overall, if I had multiple little PRs
that would have build up to such a huge change after all. Depending on the discussions and issue that arrise here, it might be beneficial to split the PR into "controversial" and "not controversial" if the "not controversial"-parts can be sensibly merged. This might ease and focus the discussion.

I'm looking forward to improve this, ping me if you need any support.

kindly,
@klees

legionth commented:
Wow thank you for your feedback! That took some time to read and evaluate :+1: 

To your questions:

> Why don't we use string for the $value in Alphanumeric? Users of Alphanumeric::getValue would either break some typehint, examine the value or use asString anyway, because they cannot possibly know, how the instance at hand was constructed. So we could store (and return) a string without any loss and remove the asString-method.

I thought about this, too and wasn't sure if this is the right intention. Because Alphnumeric values can be also integers, I wanted to create a data value store the can keep its original type, while being able to convert this value to a string.

> Why don't we introduce *Range::spans to check if a given number is within the range?

Great input I while implement it :tada:

> Why did you introduce these *AsInteger-methods (etc.) for the *Range types? Why not just use minimum and maximum and let the type hinting talk about the type?

There is no real reason behind this. I declared these names to make absolutely sure what the user expects from calling this method. I have no strong feelings about these names and can change this according to your recommendations.

> This might be the same question as above, but: Why does PositiveInteger use asInteger instead of getValue? Why does Alphanumeric have both kinds of methods?

Same as above, the names were declared like this to make absolutly clear what the user gets when calling this method.

>  Did you consciously use issue here? This would mean that we use GitHub-issues, right? In general I'm open to that, but I also think we need to consider this move carefully.

You are right this shouldn't be there. This was a text of one of my previous projects :) I will change this.

> Why do you use ilException sometimes (e.g. here)?

Because sometimes I couldn't find a native exception(https://www.php.net/manual/de/spl.exceptions.php) that would fit here, so I used default ILIAS exception.
Is there a other way you would approach this? Would you use `ConstraintViolation`?

>  Do you have any idea how the examples in files and the READMEs could be tested systematically? We have a similar challenge in the UI-framework and will have it in other places if we create more documentation for developers.

IMO the biggest challenge is to create a DIC (or the better the instances that are loaded in here). E.g. We mostly depend in our examples on an instance of `ilLanguage` because the `Validation\Factory`. If we look especially at the creation of this class we dont have the defintions of the constant `ILIAS_ABSOLUTE_PATH` or `ILIAS_ABSOLUTE_PATH`. To execute the example on console with an fresh instance, we would need to define this constant somehow.
The easiest way here to fix this without big changes would be this

```php
function __construct($a_lang_key, $absoulteILIASPath = ILIAS_ABSOLUTE_PATH)
{
    $this->lang_key = ($a_lang_key) ? $a_lang_key : $this->lang_default;
    $this->lang_path = $absoulteILIASPath . "/lang";
    $this->cust_lang_path = $absoulteILIASPath."/Customizing/global/lang";
    // ...
}
```

In the above example, we have now a optional parameter that works the same for all current consumers and opens the possibility to create own instances of this class without initialize ILIAS.
While the instantiation might work, the usage would be hard to because `ilLanguage` has hidden dependencies to `ilDBInterface`. So we would need to make this hidden dependency to injected dependency. But then we would be in need to create an instance `ilDBInterface` too. So the game starts over with `ilDB`.

Additionally I would like to mention potential security risks we could open here to making these examples work on console. Especially if we would adapt `ilDBInterface` so that it is easier to instantiate. This would mean we can manipulate the database via a script lying on the system. I don't know how high this security risk is classified or if this is risk at all (because you would still need the database user name and its password).


To your implementation request:

> I'm missing some methods on the factory with regards to the original paper that spawned this implementation: identity, allOf, byTrying, optional, customTransformation and customConstraint, as well as groups for types in data (as required/available). Is this intentional?

As outlined in the offer the following mehtods had to be implemented for the PR:

* `to` with methods `string`, `int`, `float`, `bool`, `listOf`, `dictOf`, `tupleOf`,`recordOf`, `new` and `data`
* `in` with methods `series` and `parallel`

The implementation of `kindlyTo` was an implementation apart from the offer. However if the implementation of `kindlyTo` generates to much displeasure or doubt we are willing to remove this implementation and only offer the methods mentioned in the offer.


> KindlyTo\BooleanTransformation uses a cast to bool. This might be to on the one hand. Would we really want to coerce missing values ("" or null) to false? This might be little on the other hand. Would we want to consider "true" as true but "false" as true as well?

Because this is the `KindlyTo` version we considered this was the intentional behavior to be more to the core of type casting of PHP itself. I could no found a concrete suggestion here? Is your suggestion to make this identical to `To/BooleanTransformation`? 

> KindlyTo\IntegerTransformation uses a cast to integer. Would we want to consider the float 1.999999999 to be the integer 1? I would refuse to discard that information here and simply not make the rounding operation that implicitely. We might want some rounding-transformations for that purpose. Also: I cannot see why we would want to interpret "hello" as 0, even when being kind with the incoming data.

Same as we mentioned before. The intention was to orientate on the PHP type casting. I agree we could do rounding like this but we must also consider other valid descriptions for an float e.g. the scientific way `7.16E+1`
Would the output for an integer be `71` (PHP type casting) or `72` (Our custom way). We would need first to check if this scientific description is valid and then extract if this should be round up or not. 

> To\DictionaryTransformation should not check if the transformed and the original value are the same. I currently cannot see how this needs to be different from KindlyTo\DictionaryTransformation. Same goes for To\ListTransformation and To\TupleTransformation.

We thought about this too. Our intention was that every transformation of `To` needs to be strict. E.g. `To\TupleTransformation` or `To\DictionaryTransformation` can also have `kindlyTo` transformations.
Wo considered this way:
* `To\ListTransformation` MUST return a data structures (list) of the same type
* `KindlyTo\ListTransformation` MUST return a data structure (list) but these can be of various types (Just the transformations can be executed properly).

Same goes for the other data structure transformations.

> Finally: This kind of big reviews is what you get for PRs this big =) But jokes aside: It took me some time to review this thing and write the text, but it certainly was worth it regarding the importance of the topic and the change set. I cannot see how I should have been faster overall, if I had multiple little PRs
that would have build up to such a huge change after all. Depending on the discussions and issue that arrise here, it might be beneficial to split the PR into "controversial" and "not controversial" if the "not controversial"-parts can be sensibly merged. This might ease and focus the discussion.

Normally I would prefer to take small PRs so the feedback loop is also small and we have very small self containing tasks.

```
1. Move `Transformation` to `Refinery` folder
2. Move `Validation` to `Refinery` folder.
3. Add `applyTo` method to `Transformation` interface
4. Rename Validation method `applyTo` to `restrictTo`
5. Add `Refinery` interfaces for upcoming implementation
6a. Implementation of native Datatypes as `To/Transformations`
6b. Add PositveInteger datatype
6c. Integer range implementation
6d. Add FloatRange data type
6e. Add a alphanumeric data type
7. Add new Constraint to check if the array contains only elements of the same type
8. Add DictionaryTransformation Implementation
8a. Parallel and Series Transformation implementation
....
```
These pull request would have been consecutive. The PRs with characters could have been reviewed in parallel.
Of course this appears to be more to review but I think the feedback and commenting these PRs is a whole lot easier because it is more topic specific and code is very small. Approximately these PRs can contain changes about ~150 lines. 

We will implement your suggestions now I will keep you updated for the next review round :+1:

klees committed:
Fix to-new-object example assertion

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
Fix examples return value in README

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
Remove usage

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
Remove unused comment

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
Fix exception message for StrictIntegerRange

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
Change from `owners` to `coordinators` in CONTRIBUTING.md

Co-Authored-By: legionth <niels.theen@googlemail.com>

klees committed:
At time limit in in CONTRIBUTING.md for breaking changes

Co-Authored-By: legionth <niels.theen@googlemail.com>

Commit:
Adapt change to Range classes

Commit:
Adapt examples + README to use always

Commit:
Adapt coding style + fix unit test

Commit:
Replace issue with mantis ticket

Commit:
User call_user_func_array instead of Reflection Method

Commit:
Add new datatypes to Data/Factory

Commit:
Remove unsused interfaces + Rename classes

Commit:
Use data factory for using data types from Data-library

Commit:
Fix applyTo for In/Series Transformation + Add more unit test with kindlyTo Transformations to assure this transfromation works

Commit:
Fix applyTo for In/Parallel Transformation + Add more unit test with kindlyTo Transformations to assure this transfromation works

legionth commented:
Hey @klees,

could give me some feedback of my interpretation of the following request:

> To\Group::data should return a transformation that creates a type from ILIAS\Data\, not the original factory of ILIAS\Data. That is, it should be a thin wrapper around NewMethodTransformationwhere the factory is the data-factory and the method-name needs to be passed toTo\Group::data` by the consumer.

My intentional solution for the specific method of the `To\Group` would look like this:

```php
	/**
	 * @param string $dataType - Name of the data type, this value MUST much
	 *                           with the methods provided by the `\ILIAS\Data\Factory`
	 * @return Transformation
	 * @throws \ilException
	 */
	public function data(string $dataType)
	{
		return $this->toNew(array('\ILIAS\Data\Factory', $dataType));
	}
```
So will get a implementation of a `Transformation`. Which would could be called like this:

```php
	public function testCreateTransformation()
	{
		$transformation = $this->basicGroup->data('alphanumeric');

		$this->assertInstanceOf(NewMethodTransformation::class, $transformation);

		$result = $transformation->transform(array('hello'));

		$this->assertInstanceOf(Alphanumeric::class, $result);
	}
```
Would this be consistent with your request or does this go in a completely wrong direction?
Would be very thankful if you could give me feedback on this :+1:

Commit:
Remove unused if for In/Parallel

Commit:
Rename method from asInteger to getValue

Commit:
Fix implementation of To/Group::data() and KindlyTo/Group::data() with correct behaviour according to concept paper

Commit:
Rename tests according to the original class names

Commit:
Add ConstraintViolationException to signal violation on a value

legionth commented:
> I'm currently missing ConstraintViolation (which might better be called ConstraintViolationException) in this PR (see (initial concept paper)[https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/input-processing.md]). This exception will be crucial to produce human readable error messages with i18n without binding everything to ilLng.

Reading the concept would lead us to a solution like this:

```php
class ConstraintViolationException extends \UnexpectedValueException
{
	private $translatedMessage;

	public function __construct(
		string $message,
		string $languageId,
		array $languageValues,
		\ilLanguage $language
	) {
		parent::__construct($message);

		$language->loadLanguageModule('violation');

		$this->translatedMessage = vsprintf($language->txt($languageId), $languageValues);
	}

	public function getTranslatedMessage()
	{
		return $this->translatedMessage;
	}
}
```
This would mean that the consumers like the libraries`Data` and `Validation` all need to have a dependency to `ilLanguage`. We could hide the dependencies by using just the `$DIC` in  `ConstraintViolationException::__construct(...)`. But this would mean the tests will be much more complex because we would need to mock the global `$DIC` somehow. And we all know how hard this can be.
For @mjansenDatabay and me it would be ok to pass this dependency through these classes, we just wanted to make sure that you are conform with this solution too.
Do you have any objections to pass `ilLanguage` through these classes to the `ConstraintViolationException`?

klees commented:
@legionth I would not care about `ilLanguage` here at all but instead involve it much later when we actually know whether we really need to i18n that error message, i.e. on some output layer. One of these layers would be the UI-framework (the Inputs) which has a dependency on `ilLanguage` anyway. Other layers that will want to do "human-readable" output will also already depend on `ilLanguage`. What we could offer as convience would be a method `ConstraintViolationException::getTranslatedMessage` like this:

```
public function getTranslatedMessage(callable $txt) {
    return vsprintf($txt($this->languageId), $this->languageValues);
}
```

Im working on a reply to your large initial reply atm, will be ready soon.

klees commented:
> Wow thank you for your feedback! That took some time to read and evaluate 👍

It's a pleasure =)

To your questions:

> > Why don't we use string for the $value in Alphanumeric? Users of Alphanumeric::getValue would either break some typehint, examine the value or use asString anyway, because they cannot possibly know, how the instance at hand was constructed. So we could store (and return) a string without any loss and remove the asString-method.

> I thought about this, too and wasn't sure if this is the right intention. Because Alphnumeric values can be also integers, I wanted to create a data value store the can keep its original type, while being able to convert this value to a string.

Ok, it's fine for me.

> > Why did you introduce these *AsInteger-methods (etc.) for the *Range types? Why not just use minimum and maximum and let the type hinting talk about the type?

> There is no real reason behind this. I declared these names to make absolutely sure what the user expects from calling this method. I have no strong feelings about these names and can change this according to your recommendations.

* [x] Ok. Please remove these `AsInteger`-suffixes.

> > This might be the same question as above, but: Why does PositiveInteger use asInteger instead of getValue? Why does Alphanumeric have both kinds of methods?

> Same as above, the names were declared like this to make absolutly clear what the user gets when calling this method.

* [x] Ok. Lets use `getValue` for types that only wrap other primitives.

> > Did you consciously use issue here? This would mean that we use GitHub-issues, right? In general I'm open to that, but I also think we need to consider this move carefully.

> You are right this shouldn't be there. This was a text of one of my previous projects :) I will change this.

Ok. I'm a little bit disappointed, though, because I might have hoped that you want to push this discussion =) But we have enough on the table here and I'm fine with just removing it.

> > Why do you use ilException sometimes (e.g. here)?

> Because sometimes I couldn't find a native exception(https://www.php.net/manual/de/spl.exceptions.php) that would fit here, so I used default ILIAS exception.
Is there a other way you would approach this? Would you use ConstraintViolation?

* [x] Yes, please use `ConstraintViolation` if some constraint is violated in some constructor (or elsewhere). This will be key to get error-messages that can be subject to i18n and hence be understood by ordinary users.

> The implementation of kindlyTo was an implementation apart from the offer. However if the implementation of kindlyTo generates to much displeasure or doubt we are willing to remove this implementation and only offer the methods mentioned in the offer.

* [x] I wasn't aware of that (also I certainly could have been). I would suggest to remove these methods from this PR. Please do not throw away the code, though, it's a good base for a discussion. Could you open a new PR with these transformations once this PR is merged? We could then discuss the kind transformations there.

> > KindlyTo\BooleanTransformation uses a cast to bool. This might be to on the one hand. Would we really want to coerce missing values ("" or null) to false? This might be little on the other hand. Would we want to consider "true" as true but "false" as true as well?

> Because this is the KindlyTo version we considered this was the intentional behavior to be more to the core of type casting of PHP itself. I could no found a concrete suggestion here? Is your suggestion to make this identical to To/BooleanTransformation?

Yeah, that's correct, I currently do not have a concrete suggestion =) I would not duplicate the PHP typecasting behaviour, though. It's too liberal, imo, and sometimes seems to be just wrong.

> > KindlyTo\IntegerTransformation uses a cast to integer. Would we want to consider the float 1.999999999 to be the integer 1? I would refuse to discard that information here and simply not make the rounding operation that implicitely. We might want some rounding-transformations for that purpose. Also: I cannot see why we would want to interpret "hello" as 0, even when being kind with the incoming data.

> Same as we mentioned before. The intention was to orientate on the PHP type casting. I agree we could do rounding like this but we must also consider other valid descriptions for an float e.g. the scientific way 7.16E+1
Would the output for an integer be 71 (PHP type casting) or 72 (Our custom way). We would need first to check if this scientific description is valid and then extract if this should be round up or not.

Yes. I did not think about scientific notation before, which makes things even more complicated. Another reason to discuss "kindlyTo" seperately...

> > To\DictionaryTransformation should not check if the transformed and the original value are the same. I currently cannot see how this needs to be different from KindlyTo\DictionaryTransformation. Same goes for To\ListTransformation and To\TupleTransformation.

> We thought about this too. Our intention was that every transformation of To needs to be strict. E.g. To\TupleTransformation or To\DictionaryTransformation can also have kindlyTo transformations. Wo considered this way: To\ListTransformation MUST return a data structures (list) of the same type KindlyTo\ListTransformation MUST return a data structure (list) but these can be of various types (Just the transformations can be executed properly).

I don't think that is a useful definition of strict. The strictness imo should only apply to the shape here and not to the content. Think of it this way: I know I have a record, where I absolutely want the fields, say, "firstname", "lastname" and "age", not more, not less. I would use `to->record`. There might be some wigglespace (Postels law...), however, in the way age is transfered. It might be delivered as string or integer. What I would want after my transformation, however, would be an integer. I would use `kindlyTo->int` for the "age"-field.

Using `ArrayOfSameType`, on the other hand, would forbid to, e.g., produce a list of `Result`s, since `Result` is implemented as a sum-type, where there are two distinct classes that implement one common interface. Why would we want to discard that case?

The strictness we need is on the input-side, not on the output side. The developer controls the output via the transformations anyway. 

Thanks for the feedback to the feedback, though. I'm sure it wasn't less work to digest that novel of mine and I really appreciate the interesting discussion and insights!

Commit:
Remove of KindlyTo Feature

Commit:
Replace ilException with ConstraintViolationException

Commit:
Fix docblock for ConstraintViolationException

Commit:
Remove IsArrayOfSameType Validation from datas structure transformations

Commit:
Use trait to perform applyTo on In\Group Transformations

Commit:
Add trait to reduce duplicated code in To/Group transformations

Commit:
Add trait for future implementations

Commit:
Remove usages of ArrayOfSameType

Commit:
Adapt examples because KindlyTo was removed

Commit:
Fix parameter replace ContsraintException on throwing

Commit:
Fix CodingStyle

klees commented on src/Refinery/DeriveApplyToFromTransform.php:
```diff
@@ -0,0 +1,30 @@
+			$value = $this->transform($result->value());
+		} catch (\Exception $exception) {
+			return new Result\Error($exception);
+		} catch (\Error $error) {
```
Do we really want to catch `Errors` here? This would mask every "real" error in the code-path beneath `transform`. We would only get the message, not the trace, so it would also be hard to find the problem.

klees commented on src/Refinery/DeriveTransformFromApplyTo.php:
```diff
@@ -0,0 +1,29 @@
+		/** @var Result $result */
+		$result = $this->applyTo(new Result\Ok($from));
+		if (true === $result->isError()) {
+			throw new ConstraintViolationException($result->error(), 'error');
```
`error()` could also be an exception, that we could just throw here.

Commit:
Remove catching \Error in trait

legionth commented:
I just wanted to make a quick update this PR so everybody can follow its progress :)

The general requests has been implemented, but there some things still to do. We will open several PRs that will handle different tasks:

1. A PR for the `KindlyTo` transformations. As mentioned these need some work and some discussions, we will handle this in a separated PR.
2. A PR for the movement of `Transformation` and `Validation` to move to the top level of the new Refinery. (see @klees latest comment).
3. A PR for an improved README.md. We think the current README can be used to work with this library, but there is still room for improvement.

I think we fixed the general requests for this PR. @klees and @mjansenDatabay  feel free to give me feedback on the current state :)

Commit:
Fix SplitString Unittest

Commit:
Upgrade unit tests for PHPUnit 8

legionth commented:
Removed draft state from this PR. And Rebased it according to newest trunk version.

@klees @mjansenDatabay if you are ready this PR can be merged :shipit: . I will open the next PRs when this is PR is merged, because this the base for the upcoming PRs (I want to save you from possible mail floods if you find something in this PR that needs to be changed :wink: ).

klees commented:
Hi @legionth, will do my best to give my go until the end of the week.

klees commented on src/Refinery/DeriveTransformFromApplyTo.php:
```diff
@@ -0,0 +1,37 @@
+
+			$message = $error;
+			if ($error instanceof Exception) {
+				$message = $error->getMessage();
```
```suggestion
				throw $error;
```

This might be anything, we don't really know if this is a Constraint Violation. Same for the case, where error is a `string`, but I am currently not sure, which exception we would want to use for that case. I would *not* use ConstraintViolation, because this would be caught by some mechanism later and produce a user-facing mechanism that could only say "some error". I would rather let this bubble to the outer error-catching layer, which would produce a whoops or a message via error.php.

legionth replied:
@mjansenDatabay and I discussed this and we are generally with you. Bubbling this exception up so the user has the control is a good idea in general.

We discussed this for a bit an thought of a different approach here. Instead giving the control over exceptions here we could enforce the user to use a `ConstraintViolationException` in the transformation.

In general we are using `ConstraintViolationException` in the source code of the Refinery to signal that something gone wrong in the transformation process.

What about to enforce the user to use the `ConstraintViolationException` in his custom transformation:

```php
trait DeriveTransformFromApplyTo
{
	/**
	 * @param mixed $from
	 * @return Result
	 */
	public function transform($from)
	{
		/** @var Result $result */
		$result = $this->applyTo(new Result\Ok($from));
		if (true === $result->isError()) {
			$error = $result->error();

			if ($error instanceof ConstraintViolationException) {
				throw $error;
			}

			if ($error instanceof \Exception) {
				throw new NotDefinedErrorException();  // Name not final
			}

			throw new ConstraintViolationException($error, 'error');
		}
		return $result->value();
	}
}
```
This would enforce the user to throw always a `ConstraintViolationException` instead of allowing her/him to throw other exceptions.

These were just our thoughts, but we are also fine with your solution to just throw it if its an exception :+1:

klees reviewed (changes requested):
Hi @legionth,

almost there, i think.

Almost all changes are implemented as suggested and discussed, I found one new issue (see inline comment) that needs to be solved imo.

One thing is missing, which is the naming of the ranges:

>Please indicate if the Range-types are  supposed to include or exclude the minimum and maximum values. I would further suggest to use the mathematical name Interval for these types. The question  above then could be answered by choosing one of the canonical names "open interval" or "closed interval" (or even create a type for both).

It might just be me (scientist with mathmatical background), but the "strict" for me indicated the exact opposite (like in "strictly within these borders and not on the edge"). Can we please use the mathematical names "open interval" and "closed interval" here? Or "open range" and "closed range", if interval is too mathmatical for average Joe?

Best regards!

mjansenDatabay commented:
Thx @klees 

@legionth : I'd prefer using `Open/Closed Interval` like it is already done in `\ilDateDurationInputGUI`

klees commented:
@mjansenDatabay @legionth I think we should add a `maintenance.json` indicating that this will use the coordinator-model for maintenance (with us 3 as coordinators?) and bring this to the attention of the next jour fixe. Agree?

Commit:
Fix data unit tests for PHPUnit 8

Commit:
Renamed Range and StrictRange to OpenedInterval and ClosedInterval

mjansenDatabay commented:
@klees Agree! There is already a `maintenance.json`, Niels will add a link to the file in a few minutes.

Commit:
Remove unused ilException comments

legionth commented:
@mjansenDatabay Thx for pointing it out :+1: It is hardly to find in this chaos

@klees Here you go https://github.com/ILIAS-eLearning/ILIAS/pull/1707/files#diff-1e40a9852407487bebd556260d1ba795

Commit:
Throw exception directly in "DeriveTransformFromApplyTo" + use \Exception if the result object is an string + Update README on possible used traits

Commit:
Fix Unittest of Constraint Violation for PHPUnit8

matthiaskunkel commented:
Jour Fixe, 24 APR 2019: We highly appreciate this pull request and accept it for trunk. Thanks to Michael Jansen, Niels Theen and Richard Klees for taking over the coordinatorship of the new refinery service.

Commit:
Fix newest UI components tests to match with changed namespace of refinery

Commit:
Update composer autoloader

alex40724 commented:
@mjansenDatabay I guess you already noticed, but the tests are failing.

mjansenDatabay commented:
@alex40724 Of course we did, and of course we will fix the tests.
